[RDY] options: fixes and some refactoring for number options

yagebu opened the pull request:
While trying to understand the options code for #6352 I've noticed the following issues that I'm planning to address in this PR:

1. `setlocal iminsert` and `setlocal imsearch` also set the global value.
2. For most number options `setglobal` doesn't validate the value. For example, `:setglobal sw=-4<CR>:new<CR>` crashes nvim (in Vim this is slightly more "fun" as it doesn't crash but rather allows "reverse indenting").
3. When `set <num_option>=<value>` is called with some invalid value the behaviour is very inconsistent. Depending on the option, (i) the value might be unchanged, (ii) it might be reset to the default, (iii) it might be set to the closest valid value, or (iv) it might be set to some other value (not the default). In all four cases an error message is shown and the `OptionSet` autocommand is *not* applied.

The code that handles this is currently quite a mess of validation/side-effects in no clear order. I'm planning to refactor it into validation first, side-effects second. In case the validation fails, we can reset it to the old value and return early (i.e., consistently do 3(i)).

yagebu committed:
options: setlocal should only set local value

For 'iminsert' and 'imsearch' the global value was always changed.

yagebu committed:
options: add some tests

yagebu committed:
options: move code around in set_num_option

handle side-effects after validation

yagebu committed:
options: if invalid value is given, reset to old value

justinmk commented:
`imsearch` will probably be removed, see https://github.com/neovim/neovim/issues/1708

> 1:setglobal sw=-4<CR>:new<CR>` crashes nvim

I noticed that too (with `set ts=-1`), we made `get_sw_value` and friends less forgiving. Could adjust the `assert()` or whatever else you had in mind.

> When `set <num_option>=<value>` is called with some invalid value the behaviour is very inconsistent.

Also noticed that :) Perhaps we could restore `old_value` in most cases? https://github.com/neovim/neovim/commit/7bc37ffb22a84668bba5b2e3589c4c05ad43f7d0#diff-bfacc6d8f849b9559c1c6cd5b79a5431R4206

> I'm planning to refactor it into validation first, side-effects second.

Many would welcome that. `options.lua` was a major refactor and it hasn't caused too much trouble with merging Vim patches, so cleaning up the options code is probably a net win. Adding new options is a weird experience, currently.

justinmk commented on test/functional/options/setlocal_setglobal_spec.lua:
```diff
@@ -0,0 +1,22 @@
+  helpers.clear, helpers.execute, helpers.eval, helpers.eq, helpers.nvim
+
+local function get_num_option_global(opt)
+  return nvim('command_output', 'setglobal ' .. opt .. '?'):match('%d+')
```
This is unfortunate. I noticed that `nvim_get_option` API function doesn't return the global value of a local option. Need to fix that...

justinmk commented:
instead of https://github.com/neovim/neovim/pull/6377/commits/b7dfcedd40be5b13f96e2bd12567f3dc721f43c4 can we just set `old_value` in the various cases?

justinmk commented on test/functional/options/num_options_spec.lua:
```diff
@@ -0,0 +1,42 @@
+  helpers.clear, helpers.execute, helpers.eval, helpers.eq, helpers.nvim
+
+local function get_num_option_global(opt)
+  return nvim('command_output', 'setglobal ' .. opt .. '?'):match('%d+')
```
This is unfortunate. I noticed that `nvim_get_option` API function doesn't return the global value of a local option. Need to fix that...

yagebu replied:
#6405

justinmk replied:
can use https://github.com/neovim/neovim/pull/6405 now

yagebu committed:
options: clean up num_options side-effects

yagebu committed:
options: move more validation together

yagebu committed:
options: more of the same

yagebu committed:
options: group num_option validation by type

yagebu committed:
options: fix setglobal for buf-local number options

yagebu commented:
> instead of b7dfced can we just set old_value in the various cases?

Might I ask why? If we always reset it to old_value this would add about 30 identical lines.

Any suggestions what to do with the 30 long if-else-chain? Should I put it into a separate function?

The bugs should be fixed. Will add more tests and continue with this tomorrow or in two days.

justinmk commented:
> Any suggestions what to do with the 30 long if-else-chain? Should I put it into a separate function?

It doesn't really bother me, so my suggestion is to leave it unless there is a clear improvement to be made, other than just tidyness :)

yagebu committed:
options: more tests; check first set later; stricter validation

yagebu commented:
All number options are only allowed to be between INT_MIN and INT_MAX now, which fixes the following crash (and others probably): `set sw=2147483648` (INT_MAX + 1). Allowing larger values doesn't seem to make sense for any of the options anyway.

You can still do some weird things since `vim_str2nr` simply overflows if you run it on some huge numbers. My `init.vim` now contains the following line:

    set shiftwidth=18446744073709551620  " 2^64 + 4

which is equivalent to `set sw=4` :) Not really an issue here, but that in vimscript `str2nr('4294967296') == 0` holds surprised me a bit...

Overall, I'm pretty happy with this now.

yagebu commented:
> Allowing larger values doesn't seem to make sense for any of the options anyway.

I guess it might not be if `INT_MAX` is `32767`, damn.

bfredl commented:
some discussion in #3803 whether INT_MAX is enough, it seems the conclusion at the time was that `int` is not necessarily enough, for `maxmem` for instance.

yagebu commented:
Thanks for the link. I've changed the check to allow `maxmem`, `maxmempattern`, and `maxmemtot` to be larger than INT_MAX. (and added a check for non-negativity for them, which fixes the following crash: `:set mm=-1<CR>:new<CR>`)

justinmk commented:
maxmem[tot] are going to be removed, so don't spend time on them. I'll get to that today

yagebu commented:
> maxmem[tot] are going to be removed

👍

justinmk commented on test/functional/options/num_options_spec.lua:
```diff
@@ -0,0 +1,61 @@
+    should_fail('foldlevel', -5, 'E487')
+    should_fail('foldcolumn', 100, 'E474')
+    should_fail('conceallevel', 4, 'E474')
+    should_fail('numberwidth', 20, 'E474')
```
Does this test all of the options that were re-arranged? Would feel better if each one had coverage in case of some weird side-effect.

yagebu replied:
No. Will add some more.

yagebu commented on src/nvim/option.c:
```diff
@@ -3984,235 +3984,253 @@ set_num_option (
+    if (value < 0) {
+      errmsg = e_positive;
     }
-    win_setminheight();
```
this is the only side-effect that got removed. `win_setminheight` doesn't seem to have anything to do with `winminwidth`

justinmk commented:
> Might I ask why? If we always reset it to old_value this would add about 30 identical lines.
>> I see, agreed.

Hmm. On second thought, the checks for the option type are much more verbose, and are now also duplicated. setting `oldvalue` would be less noisy than the current changes here. E.g., repeating `if (pp == &p_wh) {` is more noisy than repeating `*pp = old_value`. So I'm not sure the churn is justified.

yagebu commented:
> Hmm. On second thought, the checks for the option type are much more verbose, and are now also duplicated.

The slight duplication due to the refactoring is well worth it IMHO. Only fixing the bugs would have added plenty of checks as well, in particular since different pointer comparisons are needed for validation and side-effects in most places. Fixing the bugs while keeping my sanity and the old structure would been impossible, so I don't think this could have been done with much less churn.

The validation/side-effects split should make it much harder to reintroduce the fixed bugs compared to the old pattern of "just drop your code anywhere in here and do whatever you want".

I didn't expect this to turn up so many bugs, so this ended up being a lot more commits than I expected. I'll squash some of the later ones, but the earlier ones are hopefully split up well enough to make a review manageable.

yagebu committed:
update for changes in master; fix 'window'; tests

yagebu committed:
also test set_option

yagebu committed:
Merge branch 'master' into option-fixes

justinmk commented:
Closes https://github.com/neovim/neovim/issues/6696

yagebu committed:
Merge branch 'master' into option-fixes

yagebu committed:
update tests

yagebu commented:
This is ready again. Test failures are unrelated as far as I can tell from the output.

jamessan committed:
Merge branch 'master' into yagebu/option-fixes
